File: results/modern_tcn_metric_rebuild/53_moderntcn_fuzzyakf6d_fusion_rebuild/tools/node53_fuzzyakf6_estimator.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any

import numpy as np


def nearest_psd(p: np.ndarray, floor: float = 1e-12) -> np.ndarray:
    p = 0.5 * (p + p.T)
    values, vectors = np.linalg.eigh(p)
    return (vectors * np.maximum(values, floor)) @ vectors.T


@dataclass
class FuzzyOutput:
    theta_imu: float
    theta_acc: float
    theta_pred: float
    accel_weight: float
    accel_norm: float
    innovation: float
    nis: float
    min_covariance_eigenvalue: float
    observer_valid: bool
    gyro_energy: float
    gyro_vibration_feature: float

    @property
    def valid(self) -> bool:
        return self.observer_valid


class FuzzyAKF6D:
    """Independent Node53 copy of frozen R2_06_FuzzyAKF6D."""

    def __init__(self, cfg: dict[str, Any]):
        self.c = cfg
        self.x = np.zeros(2, dtype=float)
        self.p = np.diag([math.radians(cfg["initial_pitch_std_deg"]) ** 2, cfg["initial_bias_std_rad_s"] ** 2])
        self.initialized = False
        self.gyro_energy = 0.0

    def update(self, packet: np.ndarray) -> FuzzyOutput:
        packet = np.asarray(packet, dtype=float).reshape(-1)
        if packet.size != 6 or not np.all(np.isfinite(packet)):
            raise ValueError("Expected six finite IMU channels [fx fy fz gx gy gz]")
        accel, gyro = packet[:3], packet[3:]
        accel_norm = float(np.linalg.norm(accel))
        theta_acc = math.atan2(float(accel[0]), float(accel[2]))
        if not self.initialized:
            self.x[0] = theta_acc
            self.initialized = True
        f = np.array([[1.0, -self.c["Ts"]], [0.0, 1.0]])
        xpred = np.array([self.x[0] + self.c["Ts"] * (gyro[1] - self.x[1]), self.x[1]])
        q = np.diag([(self.c["gyro_noise_std_rad_s"] * self.c["Ts"]) ** 2, (self.c["bias_rw_std_rad_s"] * self.c["Ts"]) ** 2])
        ppred = nearest_psd(f @ self.p @ f.T + q)
        innovation = math.atan2(math.sin(theta_acc - xpred[0]), math.cos(theta_acc - xpred[0]))
        self.gyro_energy = self.c["vibration_alpha"] * self.gyro_energy + (1.0 - self.c["vibration_alpha"]) * float(gyro @ gyro)
        norm_feature = min(1.0, abs(accel_norm - self.c["g"]) / (self.c["norm_full_scale_g"] * self.c["g"]))
        vibration_feature = min(1.0, math.sqrt(max(self.gyro_energy, 0.0)) / self.c["gyro_full_scale_rad_s"])
        innovation_feature = min(1.0, abs(innovation) / math.radians(self.c["innovation_full_scale_deg"]))
        risk = self.c["norm_weight"] * norm_feature + self.c["vibration_weight"] * vibration_feature + self.c["innovation_weight"] * innovation_feature
        r = math.radians(self.c["accel_pitch_std_deg"]) ** 2 * (1.0 + self.c["max_r_multiplier"] * min(1.0, risk))
        s = float(ppred[0, 0] + r)
        k = ppred[:, 0] / s
        self.x = xpred + k * innovation
        ikh = np.eye(2) - np.outer(k, np.array([1.0, 0.0]))
        self.p = nearest_psd(ikh @ ppred @ ikh.T + np.outer(k, k) * r)
        weight = 1.0 / (1.0 + self.c["max_r_multiplier"] * min(1.0, risk))
        nis = innovation * innovation / s
        mineig = float(np.min(np.linalg.eigvalsh(self.p)))
        valid = bool(np.all(np.isfinite(self.x)) and np.all(np.isfinite(self.p)) and mineig >= -1e-10)
        return FuzzyOutput(float(self.x[0]), theta_acc, float(xpred[0]), weight, accel_norm, innovation, nis, mineig, valid, float(self.gyro_energy), float(vibration_feature))


def replay(packet: np.ndarray, cfg: dict[str, Any]) -> dict[str, np.ndarray]:
    estimator = FuzzyAKF6D(cfg)
    outputs = [estimator.update(row) for row in np.asarray(packet, dtype=float)]
    names = ("theta_imu", "theta_acc", "theta_pred", "accel_weight", "accel_norm", "innovation", "nis", "min_covariance_eigenvalue", "observer_valid", "gyro_energy", "gyro_vibration_feature")
    return {name: np.asarray([getattr(item, name) for item in outputs]) for name in names}

File: results/modern_tcn_metric_rebuild/53_moderntcn_fuzzyakf6d_fusion_rebuild/tools/node53_fuzzyakf6_estimator.m
function [state,out]=node53_fuzzyakf6_estimator(action,state_or_cfg,packet)
%NODE53_FUZZYAKF6_ESTIMATOR Independent frozen R2_06 implementation.
switch lower(char(action))
    case 'init'
        c=state_or_cfg;
        state=struct('method_id',char(c.method_id),'cfg',c,'x',[0;0], ...
            'P',diag([deg2rad(c.initial_pitch_std_deg)^2 c.initial_bias_std_rad_s^2]), ...
            'initialized',false,'gyro_energy',0);
        out=local_out(0,0,0,0,0,0,0,min(eig(state.P)),false,0,0);
    case 'update'
        state=state_or_cfg;c=state.cfg;p=double(packet(:));
        if numel(p)~=6||any(~isfinite(p));error('node53:FuzzyPacket','Expected six finite IMU channels.');end
        a=p(1:3);gyro=p(4:6);an=norm(a);theta_acc=atan2(a(1),a(3));
        if ~state.initialized;state.x(1)=theta_acc;state.initialized=true;end
        F=[1 -c.Ts;0 1];xpred=[state.x(1)+c.Ts*(gyro(2)-state.x(2));state.x(2)];
        Q=diag([(c.gyro_noise_std_rad_s*c.Ts)^2 (c.bias_rw_std_rad_s*c.Ts)^2]);
        Ppred=local_psd(F*state.P*F'+Q);innovation=atan2(sin(theta_acc-xpred(1)),cos(theta_acc-xpred(1)));
        state.gyro_energy=c.vibration_alpha*state.gyro_energy+(1-c.vibration_alpha)*(gyro'*gyro);
        fn=min(1,abs(an-c.g)/(c.norm_full_scale_g*c.g));fv=min(1,sqrt(max(state.gyro_energy,0))/c.gyro_full_scale_rad_s);fi=min(1,abs(innovation)/deg2rad(c.innovation_full_scale_deg));
        risk=c.norm_weight*fn+c.vibration_weight*fv+c.innovation_weight*fi;R=deg2rad(c.accel_pitch_std_deg)^2*(1+c.max_r_multiplier*min(1,risk));
        S=Ppred(1,1)+R;K=Ppred(:,1)/S;state.x=xpred+K*innovation;IKH=eye(2)-K*[1 0];state.P=local_psd(IKH*Ppred*IKH'+K*R*K');
        weight=1/(1+c.max_r_multiplier*min(1,risk));nis=innovation^2/S;mineig=min(eig(state.P));valid=all(isfinite(state.x))&&all(isfinite(state.P(:)))&&mineig>=-1e-10;
        out=local_out(state.x(1),theta_acc,xpred(1),weight,an,innovation,nis,mineig,valid,state.gyro_energy,fv);
    otherwise;error('node53:FuzzyAction','Unknown action.');
end
end
function out=local_out(theta,theta_acc,pred,weight,an,innovation,nis,mineig,valid,energy,vibration)
out=struct('theta_imu',double(theta),'theta_acc',double(theta_acc),'theta_pred',double(pred), ...
 'accel_weight',double(weight),'accel_norm',double(an),'innovation',double(innovation),'nis',double(nis), ...
 'NIS',double(nis),'min_covariance_eigenvalue',double(mineig),'observer_valid',logical(valid),'valid',logical(valid), ...
 'gyro_energy',double(energy),'gyro_vibration_feature',double(vibration));
end
function P=local_psd(P),P=(P+P')/2;[V,D]=eig(P);P=V*diag(max(diag(D),1e-12))*V';P=(P+P')/2;end
